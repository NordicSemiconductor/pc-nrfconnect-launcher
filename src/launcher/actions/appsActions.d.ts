/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { AnyAction } from 'redux';
import type { ThunkAction } from 'redux-thunk';

import { SourceName } from '../../ipc/sources';

declare function LoadDownloadableApps(
    appName: string,
    appSource: SourceName
): ThunkAction<void, unknown, unknown, AnyAction>;
declare function LoadDownloadableApps(): ThunkAction<
    void,
    unknown,
    unknown,
    AnyAction
>;

export const loadDownloadableApps: {
    (appName: string, appSource: SourceName): ThunkAction<
        void,
        unknown,
        unknown,
        AnyAction
    >;
    (): ThunkAction<void, unknown, unknown, AnyAction>;
};

export const setAppManagementSource: (
    source?: SourceName,
    show?: boolean
) => ThunkAction<void, unknown, unknown, AnyAction>;
