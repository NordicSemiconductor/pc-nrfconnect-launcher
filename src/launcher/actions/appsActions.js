/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { ipcRenderer, remote } from 'electron';
import { join } from 'path';
import { ErrorDialogActions } from 'pc-nrfconnect-shared';

import checkAppCompatibility from '../util/checkAppCompatibility';
import {
    EventAction,
    sendAppUsageData,
    sendLauncherUsageData,
} from './usageDataActions';

const net = remote.require('../main/net');
const fs = remote.require('fs-extra');

const mainApps = remote.require('../main/apps');
const config = remote.require('../main/config');
const settings = remote.require('../main/settings');

export const LOAD_LOCAL_APPS = 'LOAD_LOCAL_APPS';
export const LOAD_LOCAL_APPS_SUCCESS = 'LOAD_LOCAL_APPS_SUCCESS';
export const LOAD_LOCAL_APPS_ERROR = 'LOAD_LOCAL_APPS_ERROR';
export const LOAD_OFFICIAL_APPS = 'LOAD_OFFICIAL_APPS';
export const LOAD_OFFICIAL_APPS_SUCCESS = 'LOAD_OFFICIAL_APPS_SUCCESS';
export const LOAD_OFFICIAL_APPS_ERROR = 'LOAD_OFFICIAL_APPS_ERROR';
export const INSTALL_OFFICIAL_APP = 'INSTALL_OFFICIAL_APP';
export const INSTALL_OFFICIAL_APP_SUCCESS = 'INSTALL_OFFICIAL_APP_SUCCESS';
export const INSTALL_OFFICIAL_APP_ERROR = 'INSTALL_OFFICIAL_APP_ERROR';
export const REMOVE_OFFICIAL_APP = 'REMOVE_OFFICIAL_APP';
export const REMOVE_OFFICIAL_APP_SUCCESS = 'REMOVE_OFFICIAL_APP_SUCCESS';
export const REMOVE_OFFICIAL_APP_ERROR = 'REMOVE_OFFICIAL_APP_ERROR';
export const UPGRADE_OFFICIAL_APP = 'UPGRADE_OFFICIAL_APP';
export const UPGRADE_OFFICIAL_APP_SUCCESS = 'UPGRADE_OFFICIAL_APP_SUCCESS';
export const UPGRADE_OFFICIAL_APP_ERROR = 'UPGRADE_OFFICIAL_APP_ERROR';
export const SHOW_CONFIRM_LAUNCH_DIALOG = 'SHOW_CONFIRM_LAUNCH_DIALOG';
export const HIDE_CONFIRM_LAUNCH_DIALOG = 'HIDE_CONFIRM_LAUNCH_DIALOG';
export const DOWNLOAD_LATEST_APP_INFO = 'DOWNLOAD_LATEST_APP_INFO';
export const DOWNLOAD_LATEST_APP_INFO_SUCCESS =
    'DOWNLOAD_LATEST_APP_INFO_SUCCESS';
export const DOWNLOAD_LATEST_APP_INFO_ERROR = 'DOWNLOAD_LATEST_APP_INFO_ERROR';
export const SET_APP_ICON_PATH = 'SET_APP_ICON_PATH';
export const SET_APP_RELEASE_NOTE = 'SET_APP_RELEASE_NOTE';
export const SET_APP_MANAGEMENT_SHOW = 'SET_APP_MANAGEMENT_SHOW';
export const SET_APP_MANAGEMENT_FILTER = 'SET_APP_MANAGEMENT_FILTER';
export const SET_APP_MANAGEMENT_SOURCE = 'SET_APP_MANAGEMENT_SOURCE';

function loadLocalAppsAction() {
    return {
        type: LOAD_LOCAL_APPS,
    };
}

function loadLocalAppsSuccess(apps) {
    return {
        type: LOAD_LOCAL_APPS_SUCCESS,
        apps,
    };
}

function loadLocalAppsError() {
    return {
        type: LOAD_LOCAL_APPS_ERROR,
    };
}

function loadOfficialAppsAction() {
    return {
        type: LOAD_OFFICIAL_APPS,
    };
}

function loadOfficialAppsSuccess(apps, appToUpdate) {
    return {
        type: LOAD_OFFICIAL_APPS_SUCCESS,
        apps,
        appToUpdate,
    };
}

function loadOfficialAppsError() {
    return {
        type: LOAD_OFFICIAL_APPS_ERROR,
    };
}

function installOfficialAppAction(name, source) {
    return {
        type: INSTALL_OFFICIAL_APP,
        name,
        source,
    };
}

function installOfficialAppSuccessAction(name, source) {
    return {
        type: INSTALL_OFFICIAL_APP_SUCCESS,
        name,
        source,
    };
}

function installOfficialAppErrorAction() {
    return {
        type: INSTALL_OFFICIAL_APP_ERROR,
    };
}

function removeOfficialAppAction(name, source) {
    return {
        type: REMOVE_OFFICIAL_APP,
        name,
        source,
    };
}

function removeOfficialAppSuccessAction(name, source) {
    return {
        type: REMOVE_OFFICIAL_APP_SUCCESS,
        name,
        source,
    };
}

function removeOfficialAppErrorAction() {
    return {
        type: REMOVE_OFFICIAL_APP_ERROR,
    };
}

function upgradeOfficialAppAction(name, version, source) {
    return {
        type: UPGRADE_OFFICIAL_APP,
        name,
        version,
        source,
    };
}

function upgradeOfficialAppSuccessAction(name, version, source) {
    return {
        type: UPGRADE_OFFICIAL_APP_SUCCESS,
        name,
        version,
        source,
    };
}

function upgradeOfficialAppErrorAction() {
    return {
        type: UPGRADE_OFFICIAL_APP_ERROR,
    };
}

export function downloadLatestAppInfoAction() {
    return {
        type: DOWNLOAD_LATEST_APP_INFO,
    };
}

export function downloadLatestAppInfoSuccessAction() {
    return {
        type: DOWNLOAD_LATEST_APP_INFO_SUCCESS,
    };
}

export function downloadLatestAppInfoErrorAction() {
    return {
        type: DOWNLOAD_LATEST_APP_INFO_ERROR,
    };
}

export function showConfirmLaunchDialogAction(text, app) {
    return {
        type: SHOW_CONFIRM_LAUNCH_DIALOG,
        text,
        app,
    };
}

export function hideConfirmLaunchDialogAction() {
    return {
        type: HIDE_CONFIRM_LAUNCH_DIALOG,
    };
}

export function setAppIconPath(source, name, iconPath) {
    return {
        type: SET_APP_ICON_PATH,
        source,
        name,
        iconPath,
    };
}

export function setAppReleaseNoteAction(source, name, releaseNote) {
    return {
        type: SET_APP_RELEASE_NOTE,
        source,
        name,
        releaseNote,
    };
}

export function setAppManagementShow(show = {}) {
    const newState = {
        installed: true,
        available: true,
        ...settings.get('app-management.show', {}),
        ...show,
    };
    settings.set('app-management.show', newState);
    return {
        type: SET_APP_MANAGEMENT_SHOW,
        show: newState,
    };
}

export function setAppManagementFilter(filter) {
    const newState =
        filter === undefined
            ? settings.get('app-management.filter', '')
            : filter;
    settings.set('app-management.filter', newState);
    return {
        type: SET_APP_MANAGEMENT_FILTER,
        filter: newState,
    };
}

export function setAppManagementSource(source, show) {
    const sources = { ...settings.get('app-management.sources', {}) };
    if (source) {
        if (show !== undefined) {
            sources[source] = show;
        } else {
            delete sources[source];
        }
    }
    settings.set('app-management.sources', sources);
    return {
        type: SET_APP_MANAGEMENT_SOURCE,
        sources,
    };
}

/**
 * Download app icon and dispatch icon update event
 *
 * @param {string} source app source identifier
 * @param {string} name app name
 * @param {string} iconPath path to store the icon at
 * @param {string} iconUrl url to download from
 *
 * @returns {void}
 */
function downloadAppIcon(source, name, iconPath, iconUrl) {
    return dispatch => {
        // If there is a cached version, use it even before downloading.
        if (fs.existsSync(iconPath)) {
            dispatch(setAppIconPath(source, name, iconPath));
        }
        net.downloadToFile(iconUrl, iconPath, false)
            .then(() => dispatch(setAppIconPath(source, name, iconPath)))
            .catch(() => {
                /* Ignore 404 not found. */
            });
    };
}

export function loadLocalApps() {
    return dispatch => {
        dispatch(loadLocalAppsAction());
        return mainApps
            .getLocalApps()
            .then(apps => dispatch(loadLocalAppsSuccess(apps)))
            .catch(error => {
                dispatch(loadLocalAppsError());
                dispatch(ErrorDialogActions.showDialog(error.message));
            });
    };
}

export function loadOfficialApps(appName, appSource) {
    return async dispatch => {
        dispatch(loadOfficialAppsAction());
        const { fulfilled: apps, rejected: appsWithErrors } =
            await mainApps.getOfficialApps();

        dispatch(
            loadOfficialAppsSuccess(
                apps,
                appName && { name: appName, source: appSource }
            )
        );
        apps.filter(({ path }) => !path).forEach(({ source, name, url }) => {
            const iconPath = join(
                `${config.getAppsRootDir(source)}`,
                `${name}.svg`
            );
            const iconUrl = `${url}.svg`;
            dispatch(downloadAppIcon(source, name, iconPath, iconUrl));
        });
        const downloadAllReleaseNotes = (app, ...rest) => {
            if (!app) {
                return Promise.resolve();
            }
            if (
                appName &&
                !(app.name === appName && app.source === appSource)
            ) {
                return downloadAllReleaseNotes(...rest);
            }
            return mainApps
                .downloadReleaseNotes(app)
                .then(
                    releaseNote =>
                        releaseNote &&
                        dispatch(
                            setAppReleaseNoteAction(
                                app.source,
                                app.name,
                                releaseNote
                            )
                        )
                )
                .then(() => downloadAllReleaseNotes(...rest));
        };
        downloadAllReleaseNotes(...apps);

        if (appsWithErrors.length > 0) {
            handleAppsWithErrors(dispatch, appsWithErrors);
        }
    };
}

const handleAppsWithErrors = (dispatch, apps) => {
    dispatch(loadOfficialAppsError());
    apps.forEach(app => {
        dispatch(
            sendLauncherUsageData(
                EventAction.REPORT_INSTALLATION_ERROR,
                `${app.source} - ${app.name}`
            )
        );
    });

    const recover = invalidPaths => () => {
        invalidPaths.forEach(p => fs.remove(p));
        remote.getCurrentWindow().reload();
    };

    dispatch(
        ErrorDialogActions.showDialog(buildErrorMessage(apps), {
            Recover: recover(apps.map(app => app.path)),
            Close: () => dispatch(ErrorDialogActions.hideDialog()),
        })
    );
};

const buildErrorMessage = apps => {
    const errors = apps.map(app => `* \`${app.reason}\`\n\n`).join('');
    const paths = apps.map(app => `* *${app.path}*\n\n`).join('');
    return `Unable to load all apps, these are the error messages:\n\n${errors}Clicking **Recover** will attempt to remove the following broken installation directories:\n\n${paths}`;
};

export function installOfficialApp(name, source) {
    return dispatch => {
        dispatch(sendAppUsageData(EventAction.INSTALL_APP, source, name));
        dispatch(installOfficialAppAction(name, source));
        mainApps
            .installOfficialApp(name, 'latest', source)
            .then(() => {
                dispatch(installOfficialAppSuccessAction(name, source));
                dispatch(loadOfficialApps(name, source));
            })
            .catch(error => {
                dispatch(installOfficialAppErrorAction());
                dispatch(
                    ErrorDialogActions.showDialog(
                        `Unable to install: ${error.message}`
                    )
                );
            });
    };
}

export function removeOfficialApp(name, source) {
    return dispatch => {
        dispatch(sendAppUsageData(EventAction.REMOVE_APP, source, name));
        dispatch(removeOfficialAppAction(name, source));
        mainApps
            .removeOfficialApp(name, source)
            .then(() => {
                dispatch(removeOfficialAppSuccessAction(name, source));
                dispatch(loadOfficialApps(name, source));
            })
            .catch(error => {
                dispatch(removeOfficialAppErrorAction());
                dispatch(
                    ErrorDialogActions.showDialog(
                        `Unable to remove: ${error.message}`
                    )
                );
            });
    };
}

export function upgradeOfficialApp(name, version, source) {
    return dispatch => {
        dispatch(sendAppUsageData(EventAction.UPGRADE_APP, source, name));
        dispatch(upgradeOfficialAppAction(name, version, source));
        return mainApps
            .installOfficialApp(name, version, source)
            .then(() => {
                dispatch(
                    upgradeOfficialAppSuccessAction(name, version, source)
                );
                dispatch(loadOfficialApps(name, source));
            })
            .catch(error => {
                dispatch(upgradeOfficialAppErrorAction());
                dispatch(
                    ErrorDialogActions.showDialog(
                        `Unable to upgrade: ${error.message}`
                    )
                );
            });
    };
}

export function launch(app) {
    return dispatch => {
        // The apps in state are Immutable Maps which cannot be sent over IPC.
        // Converting to plain JS object before sending to the main process.
        const appObj = app.toJS();
        const sharedData =
            `App version: ${appObj.currentVersion};` +
            ` Engine version: ${appObj.engineVersion};`;
        dispatch(
            sendAppUsageData(EventAction.LAUNCH_APP, sharedData, appObj.name)
        );
        ipcRenderer.send('open-app', appObj);
    };
}

export function checkEngineAndLaunch(app) {
    return dispatch => {
        const appCompatibility = checkAppCompatibility(app);
        const launchAppWithoutWarning =
            appCompatibility.isCompatible ||
            config.isRunningLauncherFromSource();

        dispatch(
            launchAppWithoutWarning
                ? launch(app)
                : showConfirmLaunchDialogAction(
                      appCompatibility.longWarning,
                      app
                  )
        );
    };
}
