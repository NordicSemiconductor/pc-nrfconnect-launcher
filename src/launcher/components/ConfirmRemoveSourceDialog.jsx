/* Copyright (c) 2015 - 2018, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';
import { bool, func, string } from 'prop-types';

const ConfirmRemoveSourceDialog = ({
    isVisible,
    source,
    onConfirm,
    onCancel,
}) => (
    <ConfirmationDialog
        isVisible={isVisible}
        title="Remove app source"
        text={`Are you sure to remove "${source}" source along with any apps installed from it?`}
        okButtonText="Yes, remove"
        cancelButtonText="Cancel"
        onOk={() => onConfirm(source)}
        onCancel={onCancel}
    />
);

ConfirmRemoveSourceDialog.propTypes = {
    isVisible: bool.isRequired,
    source: string,
    onConfirm: func.isRequired,
    onCancel: func.isRequired,
};

ConfirmRemoveSourceDialog.defaultProps = {
    source: null,
};

export default ConfirmRemoveSourceDialog;
