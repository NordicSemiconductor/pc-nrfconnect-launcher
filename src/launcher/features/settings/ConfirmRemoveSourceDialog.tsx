/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { removeSource } from './settingsEffects';
import { getSettings, hideRemoveSourceDialog } from './settingsSlice';

export default () => {
    const { removeSource: source, isRemoveSourceDialogVisible: isVisible } =
        useLauncherSelector(getSettings);
    const dispatch = useLauncherDispatch();

    return (
        <ConfirmationDialog
            isVisible={isVisible}
            title="Remove app source"
            text={`Are you sure to remove "${source}" source along with any apps installed from it?`}
            okButtonText="Yes, remove"
            cancelButtonText="Cancel"
            onOk={() => {
                dispatch(hideRemoveSourceDialog());
                dispatch(removeSource(source!)); // eslint-disable-line @typescript-eslint/no-non-null-assertion
            }}
            onCancel={() => dispatch(hideRemoveSourceDialog())}
        />
    );
};
