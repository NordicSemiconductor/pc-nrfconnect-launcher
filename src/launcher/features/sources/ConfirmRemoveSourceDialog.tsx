/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { removeSource } from './sourcesEffects';
import {
    getIsRemoveSourceVisible,
    getSourceToRemove,
    hideRemoveSource,
} from './sourcesSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const isVisible = useLauncherSelector(getIsRemoveSourceVisible);
    const sourceToRemove = useLauncherSelector(getSourceToRemove);

    return (
        <ConfirmationDialog
            isVisible={isVisible}
            title="Remove app source"
            text={`Are you sure to remove "${sourceToRemove}" source along with any apps installed from it?`}
            okButtonText="Yes, remove"
            cancelButtonText="Cancel"
            onOk={() => {
                dispatch(hideRemoveSource());
                dispatch(removeSource(sourceToRemove!)); // eslint-disable-line @typescript-eslint/no-non-null-assertion
            }}
            onCancel={() => dispatch(hideRemoveSource())}
        />
    );
};
