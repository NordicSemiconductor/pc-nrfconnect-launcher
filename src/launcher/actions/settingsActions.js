/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { remote } from 'electron';
import { ErrorDialogActions } from 'pc-nrfconnect-shared';

import * as AppsActions from './appsActions'; // eslint-disable-line import/no-cycle

const settings = remote.require('../main/settings');
const mainApps = remote.require('../main/apps');

export const SETTINGS_LOAD = 'SETTINGS_LOAD';
export const SETTINGS_LOAD_SUCCESS = 'SETTINGS_LOAD_SUCCESS';
export const SETTINGS_LOAD_ERROR = 'SETTINGS_LOAD_ERROR';
export const SETTINGS_CHECK_UPDATES_AT_STARTUP_CHANGED =
    'SETTINGS_CHECK_UPDATES_AT_STARTUP_CHANGED';
export const SETTINGS_UPDATE_CHECK_COMPLETE_DIALOG_SHOW =
    'SETTINGS_UPDATE_CHECK_COMPLETE_DIALOG_SHOW';
export const SETTINGS_UPDATE_CHECK_COMPLETE_DIALOG_HIDE =
    'SETTINGS_UPDATE_CHECK_COMPLETE_DIALOG_HIDE';
export const SETTINGS_SOURCE_ADDED = 'SETTINGS_SOURCE_ADDED';
export const SETTINGS_SOURCE_REMOVED = 'SETTINGS_SOURCE_REMOVED';
export const SETTINGS_ADD_SOURCE_DIALOG_SHOW =
    'SETTINGS_ADD_SOURCE_DIALOG_SHOW';
export const SETTINGS_ADD_SOURCE_DIALOG_HIDE =
    'SETTINGS_ADD_SOURCE_DIALOG_HIDE';
export const SETTINGS_REMOVE_SOURCE_DIALOG_SHOW =
    'SETTINGS_REMOVE_SOURCE_DIALOG_SHOW';
export const SETTINGS_REMOVE_SOURCE_DIALOG_HIDE =
    'SETTINGS_REMOVE_SOURCE_DIALOG_HIDE';

function loadSettingsAction() {
    return {
        type: SETTINGS_LOAD,
    };
}

function loadSettingsSuccessAction(settingsObj) {
    return {
        type: SETTINGS_LOAD_SUCCESS,
        settings: settingsObj,
    };
}

function loadSettingsErrorAction(error) {
    return {
        type: SETTINGS_LOAD_ERROR,
        error,
    };
}

function checkUpdatesAtStartupChangedAction(isEnabled) {
    return {
        type: SETTINGS_CHECK_UPDATES_AT_STARTUP_CHANGED,
        isEnabled,
    };
}

export function loadSettings() {
    return dispatch => {
        dispatch(loadSettingsAction());
        try {
            let shouldCheckForUpdatesAtStartup = settings.get(
                'shouldCheckForUpdatesAtStartup'
            );
            if (shouldCheckForUpdatesAtStartup === null) {
                shouldCheckForUpdatesAtStartup = true;
            }
            const sources = settings.getSources();
            dispatch(
                loadSettingsSuccessAction({
                    shouldCheckForUpdatesAtStartup,
                    sources,
                })
            );
        } catch (error) {
            dispatch(loadSettingsErrorAction(error));
            dispatch(
                ErrorDialogActions.showDialog(
                    `Unable to load settings: ${error.message}`
                )
            );
        }
    };
}

export function checkUpdatesAtStartupChanged(isEnabled) {
    return dispatch => {
        dispatch(checkUpdatesAtStartupChangedAction(isEnabled));
        try {
            settings.set('shouldCheckForUpdatesAtStartup', isEnabled);
        } catch (error) {
            dispatch(
                ErrorDialogActions.showDialog(
                    `Unable to save settings: ${error.message}`
                )
            );
        }
    };
}

export function showUpdateCheckCompleteDialog() {
    return {
        type: SETTINGS_UPDATE_CHECK_COMPLETE_DIALOG_SHOW,
    };
}

export function hideUpdateCheckCompleteDialog() {
    return {
        type: SETTINGS_UPDATE_CHECK_COMPLETE_DIALOG_HIDE,
    };
}

function addSourceAction(name, url) {
    return {
        type: SETTINGS_SOURCE_ADDED,
        name,
        url,
    };
}

export function addSource(url) {
    return (dispatch, getState) => {
        mainApps
            .downloadAppsJsonFile(url)
            .then(source => dispatch(addSourceAction(source, url)))
            .then(() => {
                try {
                    settings.setSources(getState().settings.sources.toJS());
                } catch (error) {
                    dispatch(
                        ErrorDialogActions.showDialog(
                            `Unable to save settings: ${error.message}`
                        )
                    );
                }
            })
            .catch(error =>
                dispatch(ErrorDialogActions.showDialog(error.message))
            )
            .then(() => dispatch(AppsActions.loadOfficialApps()));
    };
}

function sourceRemovedAction(name) {
    return {
        type: SETTINGS_SOURCE_REMOVED,
        name,
    };
}

export function removeSource(name) {
    return (dispatch, getState) => {
        mainApps
            .removeSourceDirectory(name)
            .then(() => dispatch(sourceRemovedAction(name)))
            .then(() => {
                try {
                    settings.setSources(getState().settings.sources.toJS());
                } catch (error) {
                    dispatch(
                        ErrorDialogActions.showDialog(
                            `Unable to save settings: ${error.message}`
                        )
                    );
                }
            })
            .then(() => dispatch(AppsActions.loadOfficialApps()))
            .then(() => dispatch(AppsActions.setAppManagementSource(name)));
    };
}

export function showAddSourceDialog() {
    return {
        type: SETTINGS_ADD_SOURCE_DIALOG_SHOW,
    };
}

export function hideAddSourceDialog() {
    return {
        type: SETTINGS_ADD_SOURCE_DIALOG_HIDE,
    };
}

export function showRemoveSourceDialog(name) {
    return {
        type: SETTINGS_REMOVE_SOURCE_DIALOG_SHOW,
        name,
    };
}

export function hideRemoveSourceDialog() {
    return {
        type: SETTINGS_REMOVE_SOURCE_DIALOG_HIDE,
    };
}
