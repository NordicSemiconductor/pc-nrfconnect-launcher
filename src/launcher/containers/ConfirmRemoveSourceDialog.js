/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { connect } from 'react-redux';

import * as SettingsActions from '../actions/settingsActions';
import ConfirmRemoveSourceDialog from '../components/ConfirmRemoveSourceDialog';

function mapStateToProps(state) {
    return {
        source: state.settings.removeSource,
        isVisible: state.settings.isRemoveSourceDialogVisible,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        onConfirm: source => {
            dispatch(SettingsActions.hideRemoveSourceDialog());
            dispatch(SettingsActions.removeSource(source));
        },
        onCancel: () => dispatch(SettingsActions.hideRemoveSourceDialog()),
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ConfirmRemoveSourceDialog);
