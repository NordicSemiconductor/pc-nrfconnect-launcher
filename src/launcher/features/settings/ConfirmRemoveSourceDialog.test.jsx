/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import render from '../../../testrenderer';
import { showRemoveSourceDialog } from '../../actions/settingsActions';
import ConfirmRemoveSourceDialog from '../../containers/ConfirmRemoveSourceDialog';

describe('ConfirmRemoveSourceDialog', () => {
    it('is initially invisible', () => {
        expect(
            render(<ConfirmRemoveSourceDialog />).baseElement
        ).toMatchSnapshot();
    });

    it('shows the new version if one becomes available ', () => {
        expect(
            render(<ConfirmRemoveSourceDialog />, [
                showRemoveSourceDialog('the source to remove'),
            ]).baseElement
        ).toMatchSnapshot();
    });
});
