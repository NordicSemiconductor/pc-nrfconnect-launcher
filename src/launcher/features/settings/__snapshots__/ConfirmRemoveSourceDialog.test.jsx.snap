// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmRemoveSourceDialog is initially invisible 1`] = `
<body>
  <div />
</body>
`;

exports[`ConfirmRemoveSourceDialog shows the new version if one becomes available  1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-lg"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
            data-testid="title"
          >
            <h3>
              Remove app source
            </h3>
          </div>
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="modal-body"
        >
          <p
            data-testid="body"
          >
            Are you sure to remove "the source to remove" source along with any apps installed from it?
          </p>
        </div>
        <div
          class="modal-footer"
        >
           
          <button
            class="btn btn-primary"
            type="button"
          >
            Yes, remove
          </button>
          <button
            class="btn btn-outline-primary"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
